x/dao/keeper/keeper.go

```diff
@@ -0,0 +1,104 @@
+package keeper
+
+import (
+	"fmt"
+
+	"cosmossdk.io/math"
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/cosmos/cosmos-sdk/codec"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	disttypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+
+	"planetmint-go/config"
+	"planetmint-go/x/dao/types"
+)
+
+type (
+	Keeper struct {
+		cdc        codec.BinaryCodec
+		storeKey   storetypes.StoreKey
+		memKey     storetypes.StoreKey
+		paramstore paramtypes.Subspace
+
+		bankKeeper    types.BankKeeper
+		accountKeeper types.AccountKeeper
+	}
+)
+
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey,
+	memKey storetypes.StoreKey,
+	ps paramtypes.Subspace,
+
+	bankKeeper types.BankKeeper,
+	accountKeeper types.AccountKeeper,
+) *Keeper {
+	// set KeyTable if it has not already been set
+	if !ps.HasKeyTable() {
+		ps = ps.WithKeyTable(types.ParamKeyTable())
+	}
+
+	return &Keeper{
+		cdc:        cdc,
+		storeKey:   storeKey,
+		memKey:     memKey,
+		paramstore: ps,
+
+		bankKeeper:    bankKeeper,
+		accountKeeper: accountKeeper,
+	}
+}
+
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+func (k Keeper) DistributeCollectedFees(ctx sdk.Context) {
+	ctx = sdk.UnwrapSDKContext(ctx)
+	conf := config.GetConfig()
+
+	balances := make(map[string]math.Int)
+	totalStake := math.ZeroInt()
+	k.accountKeeper.IterateAccounts(ctx, func(acc authtypes.AccountI) bool {
+		addr := acc.GetAddress()
+		balance := k.bankKeeper.SpendableCoins(ctx, addr)
+		found, stake := balance.Find(conf.StakeDenom)
+		if found {
+			totalStake = totalStake.Add(stake.Amount)
+			balances[addr.String()] = stake.Amount
+		}
+		return false
+	})
+
+	distAddr := k.accountKeeper.GetModuleAddress(disttypes.ModuleName)
+	distSpendableCoins := k.bankKeeper.SpendableCoins(ctx, distAddr)
+	found, coinToDistribute := distSpendableCoins.Find(conf.FeeDenom)
+
+	if found {
+		decTotalAmountToDistribute := sdk.NewDecFromInt(coinToDistribute.Amount)
+		decTotalStake := sdk.NewDecFromInt(totalStake)
+		for addr, stake := range balances {
+			decStake := sdk.NewDecFromInt(stake)
+			share := decStake.Quo(decTotalStake)
+			claim := decTotalAmountToDistribute.Mul(share)
+			if claim.GTE(sdk.OneDec()) {
+				intClaim := claim.TruncateInt()
+				coinClaim := sdk.NewCoin(conf.FeeDenom, intClaim)
+				accAddress, err := sdk.AccAddressFromBech32(addr)
+				if err != nil {
+					panic(err)
+				}
+				if !k.bankKeeper.BlockedAddr(accAddress) {
+					err = k.bankKeeper.SendCoinsFromModuleToAccount(ctx, disttypes.ModuleName, accAddress, sdk.NewCoins(coinClaim))
+					if err != nil {
+						panic(err)
+					}
+				}
+			}
+		}
+	}
+}
```

Comment:
this If statement is very complex and hard to read could you split it in to smaller functions with proper names?

Something like this:

```
if found {
	err := processBalances(balances, totalStake, coinToDistribute, k, ctx)
	if err != nil {
		// Handle the error, perhaps by logging it or returning it further up
		fmt.Println("Error processing balances:", err)
	}
}
```

```
// Check if the address is blocked
func isAddressBlocked(k *YourType, accAddress sdk.AccAddress) bool {
	return k.bankKeeper.BlockedAddr(accAddress)
}

// Send coins from the module to the account
func sendCoinsFromModuleToAccount(k *YourType, ctx *YourContextType, accAddress sdk.AccAddress, coinClaim sdk.Coin) error {
	return k.bankKeeper.SendCoinsFromModuleToAccount(ctx, disttypes.ModuleName, accAddress, sdk.NewCoins(coinClaim))
}

// Calculate the claim for an address
func calculateClaimForAddress(stake sdk.Int, totalStake sdk.Int, coinToDistribute sdk.Coin) sdk.Dec {
	decTotalAmountToDistribute := sdk.NewDecFromInt(coinToDistribute.Amount)
	decTotalStake := sdk.NewDecFromInt(totalStake)
	decStake := sdk.NewDecFromInt(stake)

	share := decStake.Quo(decTotalStake)
	return decTotalAmountToDistribute.Mul(share)
}

func processBalances(balances map[string]sdk.Int, totalStake sdk.Int, coinToDistribute sdk.Coin, k *YourType, ctx *YourContextType) error {
	for addr, stake := range balances {
		claim := calculateClaimForAddress(stake, totalStake, coinToDistribute)

		if claim.GTE(sdk.OneDec()) {
			intClaim := claim.TruncateInt()
			coinClaim := sdk.NewCoin(conf.FeeDenom, intClaim)

			accAddress, err := sdk.AccAddressFromBech32(addr)
			if err != nil {
				return err
			}

			if !isAddressBlocked(k, accAddress) {
				err = sendCoinsFromModuleToAccount(k, ctx, accAddress, coinClaim)
				if err != nil {
					return err
				}
			}
		}
	}
	return nil
}
```